src/Numerics/SoaSphericalTensor.h

```diff
@@ -55,21 +59,58 @@ struct SoaSphericalTensor
   SoaSphericalTensor(const SoaSphericalTensor& rhs) = default;
 
   ///compute Ylm
-  void evaluate_bare(T x, T y, T z, T* Ylm) const;
+  static void evaluate_bare(T x, T y, T z, T* Ylm, int lmax, const T* factorL, const T* factorLM);
 
   ///compute Ylm
   inline void evaluateV(T x, T y, T z, T* Ylm) const
   {
-    evaluate_bare(x, y, z, Ylm);
+    evaluate_bare(x, y, z, Ylm, Lmax, FactorL.data(), FactorLM.data());
     for (int i = 0, nl = cYlm.size(); i < nl; i++)
       Ylm[i] *= NormFactor[i];
   }
 
+  /**
+   * @brief evaluate V for multiple electrons and multiple pbc images
+   * 
+   * @param [in] xyz electron positions [Nelec, Npbc, 3(x,y,z)]
+   * @param [out] Ylm Spherical tensor elements [Nelec, Npbc, Nlm]
+  */
+  inline void batched_evaluateV(const OffloadArray3D& xyz, OffloadArray3D& Ylm) const
+  {
+    const size_t nElec = xyz.size(0);
+    const size_t Npbc  = xyz.size(1); // number of PBC images
+    assert(xyz.size(2) == 3);
+
+    assert(Ylm.size(0) == nElec);
+    assert(Ylm.size(1) == Npbc);
+    const size_t Nlm = Ylm.size(2);
+
+    size_t nR = nElec * Npbc; // total number of positions to evaluate
+
+    auto* xyz_ptr        = xyz.data();
+    auto* Ylm_ptr        = Ylm.data();
+    auto* FactorLM_ptr   = FactorLM.data();
+    auto* FactorL_ptr    = FactorL.data();
+    auto* NormFactor_ptr = NormFactor.data();
+
+
+    PRAGMA_OFFLOAD("omp target teams distribute parallel for \
```

Comment:
Could you add a FIXME saying the "always" should be removed after making the transfer only once ahead of time?

Reply:
fixed